sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala

```diff
@@ -78,96 +77,265 @@ abstract class GpuWindowInPandasExecMetaBase(
+      }
+    }.map(GpuColumnVector.from(_, IntegerType))
+    new ColumnarBatch(boundsCVs.toArray, batch.numRows())
+  }
+
+  /*
+   * (Currently do it be leveraging the existing APIs in cuDF Java.
+   *  But still need a dedicated API for better performance.)
+   */
+  private[this] def genGroupInfoFromBatch(batch: ColumnarBatch): Seq[Int] = {
+    assert(batch != null, "batch is null.")
+    if (batch.numRows() > 0 && batch.numCols() > 0 && partitionRefs.nonEmpty) {
+      //  a) Run count aggregation on each group in the batch, including null values.
+      //  b) Restore the original order (Ascending with NullsFirst) defined in window plan,
+      //     since "cudf.Table.groupBy" will break the order.
+      //  c) Copy the 'count' column to host and it is the group info.
+      val partitionIndices = partitionRefs.indices
+      val cntTable = withResource(GpuProjectExec.project(batch, partitionRefs)) { projected =>
+        withResource(GpuColumnVector.from(projected)) { table =>
+          table
+            .groupBy(partitionIndices:_*)
+            .aggregate(cudf.Aggregation.count(cudf.Aggregation.NullPolicy.INCLUDE).onColumn(0))
         }
-        val splitIndices = withResource(countHostCol) { cntCol =>
-          // Verified the type of Count column is integer, so use getInt
-          // Also drop the last count value due to the split API definition.
-          (0L until (numRows - 1)).map(id => cntCol.getInt(id))
-        }.scan(0)(_+_).tail
-
-        // 2) Split the table when needed
-        val resultBatch = if (splitIndices.nonEmpty) {
-          // More than one group, split it and close this incoming batch
-          withResource(batch) { cb =>
-            withResource(GpuColumnVector.from(cb)) { table =>
-              withResource(table.contiguousSplit(splitIndices:_*)) { tables =>
-                // Return the first one and enqueue others
-                val splitBatches = tables.safeMap(table =>
-                  GpuColumnVectorFromBuffer.from(table, GpuColumnVector.extractTypes(batch)))
-                groupBatches.enqueue(splitBatches.tail.map(sb =>
-                  SpillableColumnarBatch(sb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY,
-                    spillCallback)): _*)
-                splitBatches.head
-              }
+      }
+      val orderedTable = withResource(cntTable) { table =>
+        table.orderBy(partitionIndices.map(id => cudf.Table.asc(id, true)): _*)
+      }
+      val (countHostCol, numRows) = withResource(orderedTable) { table =>
+        // Yes copying the data to host, it would be OK since just copying the aggregated
+        // count column.
+        (table.getColumn(table.getNumberOfColumns - 1).copyToHost(), table.getRowCount)
+      }
+      withResource(countHostCol) { cntCol =>
+        // Verified the type of Count column is integer, so use getInt
+        (0L until numRows).map(id => cntCol.getInt(id))
```

Comment:
Is this lazy? I get nervous around `map` because in some cases it will not do the calculation now, it will just capture the map function and do it later on.  If that is happening here, then the host column may have been released by the time this code executes.